### PR TITLE
feat: join subgroup FS-1796 FS-1795, FS-1841

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-1796" title="FS-1796" target="_blank">TaskFS-1796</a>  [iOS] Support joining subgroups
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
In this PR we add support joining a subgroup. which either entails creating a new subgroup, deleting a stale subgroup then re-creating it, or joining an existing valid subgroup.

#### Test Coverage

- Unit test asserting creating a new subgroup
- Unit test asserting deleting a stale subgroup and recreating it
- Unit test asserting joining an existing subgroup
- Unit tests for deleting subgroup action

### Notes 

To help with the testing, I've moved away from some manual mocks and use the auto-generated mocks instead.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by David-Henner — APPROVED** [review]

**Review by agisilaos — APPROVED** [review]

File: wire-ios-data-model/Tests/MLS/MLSServiceTests.swift [hunk 1]
```diff
@@ -1722,3 +1891,30 @@ extension MLSGroupID {
...
+
+    static func random(length: UInt) -> String {
+        let randomChars = (0..<length).compactMap { _ in
+            "a...z".randomElement()
+        }
+
+        return String(randomChars)
+    }
+
+}
```
Comment (netbe) [inline wire-ios-data-model/Tests/MLS/MLSServiceTests.swift]: is this only used here ? If so consider use private or move to some utils files instead

File: wire-ios-data-model/Tests/MLS/MockMLSService.swift [hunk 2]
```diff
@@ -172,4 +172,11 @@ class MockMLSService: MLSServiceInterface {
         calls.scheduleCommitPendingProposals.append((groupID, commitDate))
     }
 
+    func createOrJoinSubgroup(
+        parentQualifiedID: QualifiedID,
+        parentID: MLSGroupID
+    ) async {
+
```
Comment (netbe) [inline wire-ios-data-model/Tests/MLS/MockMLSService.swift:179]: finish mock implementation or fatalError()

File: wire-ios-request-strategy/Tests/Sources/Mocks/MockMLSService.swift [hunk 3]
```diff
@@ -105,4 +105,15 @@ class MockMLSService: MLSServiceInterface {
...
+        parentQualifiedID: QualifiedID,
+        parentID: MLSGroupID
+    ) async {
+
+    }
+
+    func generateConferenceInfo(for groupID: MLSGroupID) throws -> MLSConferenceInfo {
+        fatalError("not implemented")
+    }
 }
```
Comment (netbe) [inline wire-ios-request-strategy/Tests/Sources/Mocks/MockMLSService.swift:119]: Does this still make sense to have different implementations of same mocks across different frameworks?

File: wire-ios-data-model/Source/MLS/MLSActionsProvider.swift [hunk 4]
```diff
@@ -55,9 +56,24 @@ protocol MLSActionsProviderProtocol {
...
+    func deleteSubgroup(
+        conversationID: UUID,
+        domain: String,
+        subgroupType: SubgroupType,
+        context: NotificationContext
+    ) async throws
+
 }
 
 class MLSActionsProvider: MLSActionsProviderProtocol {
```
Comment (netbe) [inline wire-ios-data-model/Source/MLS/MLSActionsProvider.swift]:
```suggestion
final class MLSActionsProvider: MLSActionsProviderProtocol {
```

**Review by netbe — COMMENTED** [review]
LGTM;) just some thoughts

File: wire-ios-request-strategy/Tests/Sources/Mocks/MockMLSService.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Tests/Sources/Mocks/MockMLSService.swift:119]: I want to generate this mock and reuse it but I thought to do it in another PR.

File: wire-ios-data-model/Tests/MLS/MockMLSService.swift (hunk 2 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Tests/MLS/MockMLSService.swift:179]: I want to generate this mock and reuse it but I thought to do it in another PR.

File: wire-ios-data-model/Tests/MLS/MLSServiceTests.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Tests/MLS/MLSServiceTests.swift]: I moved them to separate files

File: wire-ios-data-model/Source/MLS/MLSActionsProvider.swift (hunk 4 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/MLS/MLSActionsProvider.swift]: Fixed

**Review by netbe — APPROVED** [review]

**Pull request merged**